jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/ContentSourcePublisher.java

```diff
@@ -13,147 +13,311 @@
  * ultimately call {@link Flow.Subscriber#onNext(Object)} once content is available.</p>
+ * <p>{@link Content.Source} can be consumed only once and does not support multicast subscription.
+ * {@link Content.Source} will be consumed fully, otherwise will be failed in case of any errors
+ * to prevent resource leaks.</p>
  */
 public class ContentSourcePublisher implements Flow.Publisher<Content.Chunk>
 {
-    private final Content.Source content;
+    private static final Logger LOG = LoggerFactory.getLogger(ContentSourcePublisher.class);
+
+    private final AtomicReference<Content.Source> content;
 
     public ContentSourcePublisher(Content.Source content)
     {
-        this.content = content;
+        if (content == null)
+            throw new IllegalArgumentException("Content.Source must not be null");
+        this.content = new AtomicReference<>(content);
     }
 
     @Override
     public void subscribe(Flow.Subscriber<? super Content.Chunk> subscriber)
     {
-        subscriber.onSubscribe(new SubscriptionImpl(content, subscriber));
+        // As per rule 1.11, we have decided to support SINGLE subscriber
+        // in a UNICAST configuration for this implementation. It means
+        // that Content.Source can be consumed only once.
+        Content.Source content = this.content.getAndSet(null);
+        if (content != null)
+            onSubscribe(subscriber, content);
+        else
+            onMultiSubscribe(subscriber);
     }
 
-    private static class SubscriptionImpl implements Flow.Subscription
+    private void onSubscribe(Flow.Subscriber<? super Content.Chunk> subscriber, Content.Source content)
     {
-        private final AutoLock lock = new AutoLock();
-        private final Content.Source content;
-        private final Flow.Subscriber<? super Content.Chunk> subscriber;
-        private long demand;
-        private boolean stalled;
-        private boolean cancelled;
-        private boolean terminated;
+        // As per rule 1.9, we need to throw a `java.lang.NullPointerException`
+        // if the `Subscriber` is `null`
+        if (subscriber == null)
+        {
+            NullPointerException error = new NullPointerException("Flow.Subscriber must not be null");
+            content.fail(error);
+            throw error;
+        }
+        LastWillSubscription subscription = new ActiveSubscription(content, subscriber);
+        // As per rule 1.9, this method must return normally (i.e. not throw).
+        try
+        {
+            subscriber.onSubscribe(subscription);
+        }
+        catch (Throwable err)
+        {
+            // As per rule 2.13, we MUST consider subscription cancelled and
+            // MUST raise this error condition in a fashion that is adequate for the runtime environment.
+            subscription.cancel(err, LastWillSubscription.FinalSignal.SUPPRESS);
+            LOG.error("Flow.Subscriber " + subscriber + " violated rule 2.13", err);
```

Comment:
Wouldn't it be better to rethrow rather than logging?

Reply:
@lorban According to the specification we cannot throw any exception
1. As per rule 1.9, subscribe method must return normally (i.e. not throw)
2. As per rule 2.13, ...the caller MUST raise this error condition in a fashion that is adequate for the runtime environment. «Raise this error condition in a fashion that is adequate for the runtime environment» could mean logging the error—or otherwise make someone or something aware of the situation—as the error cannot be signalled to the faulty Subscriber.